lib/decoder_impl.cc

```diff
@@ -19,804 +19,1021 @@
+                    this->d_debug_samples.write(reinterpret_cast<const char *>(&v[i]), sizeof(gr_complex));
+                }
+            #else
+                (void) v;
+                (void) length;
+            #endif
+        }
+
+        bool decoder_impl::calc_energy_threshold(gr_complex *samples, uint32_t window_size, float threshold) {
+            float result = 0.0f;
 
-        instantaneous_frequency(samples, samples_ifreq, window);
-        return norm_cross_correlate(samples_ifreq, &d_downchirp_ifreq[0], window);
-    }
+            for (uint32_t i = 0u; i < window_size; i++) {
+                const float magn = abs(samples[i]);
+                result += magn * magn;
+            }
 
-    /**
-     * Calculate normalized cross correlation of real values.
-     * See https://en.wikipedia.org/wiki/Cross-correlation#Normalized_cross-correlation.
-     */
-    float decoder_impl::norm_cross_correlate(const float *samples_1, const float *samples_2, uint32_t window) {
-        float result = 0.0f;
+            result /= (float)window_size;
 
-        double average_1 = std::accumulate(samples_1, samples_1 + window, 0.0) / window;
-        double average_2 = std::accumulate(samples_2, samples_2 + window, 0.0) / window;
-        double sd_1 = stddev(samples_1, window, average_1);
-        double sd_2 = stddev(samples_2, window, average_2);
+            #ifndef NDEBUG
+                this->d_debug << "T: " << result << "\n";
+            #endif
 
-        for (int i = 0; i < window-1; i++) {
-            result += (samples_1[i] - average_1) * (samples_2[i] - average_2) / (sd_1 * sd_2);
+            return result > threshold;
         }
-        result = result / (window-1);
 
-        return result;
-    }
+        void decoder_impl::instantaneous_frequency(const gr_complex *in_samples, float *out_ifreq, uint32_t window) {
+            if (window < 2) {
+                // TODO: throw warning here
+                std::cerr << "LoRa Decoder Warning: window size < 2 !" << std::endl;
+                return;
+            }
+
+            /* instantaneous_phase */
+            float iphase_1, iphase_2;
 
-    float decoder_impl::sliding_norm_cross_correlate(const float *samples_1, const float *samples_2, uint32_t window, uint32_t slide, int32_t* index) {
-        float correlations[slide*2];
-        float samples_1_padded[window+slide*2];
+            for (uint32_t i = 1; i < window; i++) {
+                iphase_1 = std::arg(in_samples[i - 1]);
+                iphase_2 = std::arg(in_samples[i]);
 
-        double average_1 = std::accumulate(samples_1, samples_1 + window, 0.0) / window;
-        double average_2 = std::accumulate(samples_2, samples_2 + window, 0.0) / window;
-        double sd_1 = stddev(samples_1, window, average_1);
-        double sd_2 = stddev(samples_2, window, average_2);
+                // Unwrapped loops from liquid_unwrap_phase
+                while ( (iphase_2 - iphase_1) >  M_PI ) iphase_2 -= 2.0f*M_PI;
```

Comment:
Does this offer a significant performance advantage? If not, I would prefer calling liquid_unwrap_phase.